e2e/LongHaul/service/IotHub.cs

```diff
@@ -91,12 +97,79 @@ public async Task SetDesiredPropertiesAsync(string keyName, string properties, C
 
                 await Task.Delay(s_desiredPropertiesSetInterval, ct).ConfigureAwait(false);
             }
         }
 
+        public async Task SendC2dMessagesAsync(CancellationToken ct)
+        {
+            await s_serviceClient.Messages.OpenAsync(ct).ConfigureAwait(false);
+
+            while (!ct.IsCancellationRequested)
+            {
+                var payload = new CustomC2dMessagePayload
+                {
+                    RandomId = Guid.NewGuid(),
+                    CurrentTimeUtc = DateTime.UtcNow,
+                    MessagesSentCount = ++_totalC2dMessagesSentCount,
+                };
+                var message = new OutgoingMessage(payload)
+                {
+                    // An acknowledgment is sent on delivery success or failure.
+                    Ack = DeliveryAcknowledgement.Full,
+                    MessageId = payload.RandomId.ToString(),
+                };
+
+                _logger.Trace($"Sending message with Id {message.MessageId} to the device: {_deviceId}", TraceSeverity.Information);
+                _logger.Metric(TotalC2dMessagesSentCount, _totalC2dMessagesSentCount);
+
+                await s_serviceClient.Messages.SendAsync(_deviceId, message, ct).ConfigureAwait(false);
+
+                await Task.Delay(s_c2dMessagesSentInterval, ct).ConfigureAwait(false);
+            }
+
+            await s_serviceClient.Messages.CloseAsync(ct).ConfigureAwait(false);
+        }
+
+        public async Task ReceiveMessageFeedbacksAsync(CancellationToken ct)
+        {
+            // It is important to note that receiver only gets feedback messages when the device is actively running and acting on messages.
+            _logger.Trace("Starting to listen to cloud-to-device feedback messages", TraceSeverity.Verbose);
+
+            AcknowledgementType OnC2dMessageAck(FeedbackBatch feedbackMessages)
+            {
+                foreach (FeedbackRecord feedbackRecord in feedbackMessages.Records)
+                {
+                    _logger.Trace(
+                        $"Device {feedbackRecord.DeviceId} acted on message: {feedbackRecord.OriginalMessageId} with status: {feedbackRecord.StatusCode}",
+                        TraceSeverity.Information);
+                }
+
+                _totalFeedbackMessagesReceivedCount += feedbackMessages.Records.Count();
+                _logger.Metric(TotalFeedbackMessagesReceivedCount, _totalFeedbackMessagesReceivedCount);
+
+                return AcknowledgementType.Complete;
+            }
+
+            s_serviceClient.MessageFeedback.MessageFeedbackProcessor = OnC2dMessageAck;
+            await s_serviceClient.MessageFeedback.OpenAsync(ct).ConfigureAwait(false);
+
+            try
+            {
+                await Task.Delay(-1, ct);
+            }
+            catch (OperationCanceledException) { }
+
+            await s_serviceClient.MessageFeedback.CloseAsync(ct).ConfigureAwait(false);
```

Comment:
finally

Reply:
Oh, don't pass the ct in, because we quit when it is signaled so we'll fail to close here.

Either use no cancellation token, or create one for like 5 seconds or something.